### ISSUE TITLE
Sometimes websocket session "silently dies" causing CCS fail to deliver commands

### ISSUE THREAD

**Issue by calmnesss** [openremote/CCS#19]
We observed that after some time commands are failed to be delivered. We collected all logs we have and analyzed them. Here is basically step by step analyzis:

Other issues seems to be caused by bug on beehive side at least that's how it looks like. For example we could look into 6186194181 account - it failed to send a command around 22:10:

`message:2017-06-28T22:08:59,439 [pool-6-thread-327] ERROR ControllerProxyFacade:108 - ControllerCmd{deviceName='0000000000', commandName='UPDATE_DEVICE_DATA', commandParam='{"reason":"ModeChanged","id":690,"name":"Motion Sensor 2","beehiveName":"027ba01cf5","type":"MOTION_DETECTOR","model":"dectMotionDetector","baseAnnouncementCommand":"027ba01cf5-Announcement-U1","baseAnnouncementState":"off","ceasedDelay":"-1","raisedDelay":"0"}', username='752mvh7te', type=CONTROLLER_MANAGE_DEVICE_DATA_CMD} send failed: Unable to invoke request SPN-6186194181 transaction-5aaa1c48-1ec7-4cee-8d5f-58347405c177`

Attempt to send this command was started at 22:08:39 (and since it failed at 22:08:59 it failed by timeout - timeout is 20 sec)

`message:2017-06-28T22:08:39,388 [pool-6-thread-327] INFO ControllerProxyFacade:76 - Sending ControllerCmd{deviceName='0000000000', commandName='UPDATE_DEVICE_DATA', commandParam='{"reason":"ModeChanged","id":690,"name":"Motion Sensor 2","beehiveName":"027ba01cf5","type":"MOTION_DETECTOR","model":"dectMotionDetector","baseAnnouncementCommand":"027ba01cf5-Announcement-U1","baseAnnouncementState":"off","ceasedDelay":"-1","raisedDelay":"0"}', username='752mvh7te', type=CONTROLLER_MANAGE_DEVICE_DATA_CMD} SPN-6186194181 transaction-5aaa1c48-1ec7-4cee-8d5f-58347405c177 type:hms offset:402,419,313`

I checked that no other account has tried to send any commands at this time moment and went to beehive logs. Beehive logs are following:

`source:/srv/logs/beehive/ccs.log message:Jun 28 22:08:39 beehive1-eqix-sv5 [BeehiveControllerProxy] [Thread-1619404] INFO ProxyServer:358 - Proxy command could not be send type:beehive offset:3,783,715 input_type:log tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:June 28th 2017, 22:08:39.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5 beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AVzwv-9FtjMqAG1LqhZG _type:beehive _index:logstash-sv5-2017.06.28 _score: -`

This log message is written for only following cases: either CCS doesn't have websocket connection for that account, either it is considered "not-open" -> CCS closes the connection and prints that log message. At this case no web socket notification is send to Controller thus it doesn't pickup the command.

I checked both Controller and HMS logs trying to find out if Controller was offline at the moment, but haven't found any log messages which would prove that point:
 - no online/offline messages on HMS side, not websocket connection was closed messages on Controller side around that time
- last websocket connection reestablishment happened on June 26 around 11:46 and seems was caused by connectivity issues to backend
- last known online/offline state change which was detected by backend happened on June 20

Last successful command execution happened for that account was on June 21:

`message:2017-06-21T22:05:37,090 [pool-6-thread-479] INFO ControllerProxyFacade:102 - ControllerCmd{deviceName='0000000000', commandName='UPDATE_DEVICE_DATA', commandParam='{"reason":"ModeChanged","id":690,"name":"Motion Sensor 2","beehiveName":"027ba01cf5","type":"MOTION_DETECTOR","model":"dectMotionDetector","baseAnnouncementCommand":"027ba01cf5-Announcement-U1","baseAnnouncementState":"off","ceasedDelay":"-1","raisedDelay":"0"}', username='752mvh7te', type=CONTROLLER_MANAGE_DEVICE_DATA_CMD} is successfully posted SPN-6186194181 transaction-cb75df91-e7a9-4e9c-84c4-17597083cb76 `

I may try to look into other accounts which had similar beehive messages just to try to see a pattern, but for now it looks like WebSocket connection "silently dies" while Controller fails to detect that and continues to believe that it has open websocket connection.

# Some more info:

For period from 21 June till today first failed command happened on June 26 at 12:35

`message:2017-06-26T12:35:00,569 [pool-6-thread-472] ERROR ControllerProxyFacade:108 - ControllerCmd{deviceName='0000000000', commandName='UPDATE_DEVICE_DATA', commandParam='{"reason":"ModeChanged","id":689,"name":"Motion Sensor 1","beehiveName":"027ba01c0d","type":"MOTION_DETECTOR","model":"dectMotionDetector","baseAnnouncementCommand":"027ba01c0d-Announcement-U1","baseAnnouncementState":"off","ceasedDelay":"-1","raisedDelay":"0"}', username='752mvh7te', type=CONTROLLER_MANAGE_DEVICE_DATA_CMD} send failed: Unable to invoke request SPN-6186194181 transaction-4b582c18-646f-45b2-b116-249e6c1335ba`

Last successful attempt happened on June 21:

`message:2017-06-21T22:05:35,241 [pool-6-thread-479] INFO ControllerProxyFacade:76 - Sending ControllerCmd{deviceName='0000000000', commandName='UPDATE_DEVICE_DATA', commandParam='{"reason":"ModeChanged","id":690,"name":"Motion Sensor 2","beehiveName":"027ba01cf5","type":"MOTION_DETECTOR","model":"dectMotionDetector","baseAnnouncementCommand":"027ba01cf5-Announcement-U1","baseAnnouncementState":"off","ceasedDelay":"-1","raisedDelay":"0"}', username='752mvh7te', type=CONTROLLER_MANAGE_DEVICE_DATA_CMD} SPN-6186194181 transaction-cb75df91-e7a9-4e9c-84c4-17597083cb76 type:hms SPN:6186194181`

After that no more attempts were performed until June 26 - since then all of them started failing.

**Comment by calmnesss** [comment]
I will put new information as soon as I collect it.

At the mean time I looked into the CCS code and found one potentially dangerous place which could lead to undefined behavior inside ControllerSessionHandler class:

`private final Map<String,Session> sessions = new HashMap<>();`

sessions object doesn't use concurrent implementation of HashMap and is not implicitly synchronized from what I can see, however it is accessed and modified by multiple servlet threads. It very might be the cause of the issue above.

**One slight note though regarding beehive logs**

From ControllerSessionHandler:
```
    public void sendToController(String username, JSONObject message) throws WSException {

        Session session = sessions.get(username);
        if (session != null && session.isOpen()) {
            sendToSession(session, message);
        } else {
            if (session != null) {
                removeSession(session);
            }
            throw new WSException("No session for user : " + username);
        }
    }
```

Information regarding which user doesn't have a session is available within an exception, however it all lost when WSException is catched within ProxyClient class:

```
 catch (WSException e) {
             logger.info("Proxy command could not be send");
          }
```

This creates difficulties when analyzing logs especially if series of commands from different users were executed at the moment of time. Could it be changed the way that this information is not lost?

**Comment by mickz18** [comment]
Since the implementation  of WSHOOK I have put the sessions to a ConcurrentHashMap this could solve
the problem, I have change the log to have more info

**Comment by calmnesss** [comment]
Ok, but it is going to take time before we can say if it helped to solve the issue.

Also spent some more time on this looking into other accounts: it seems to start happening after websocket connection was recreated from Controller side - after it got recreated commands just are not passing through with the symptomes mentioned above: ccs doesn't have a session while Controller thinks it is ok. It is not happening for every connection recreation, just for some. Unfortunately nothing more can be claimed since adding/removing web socket session is not logged at all on ccs level.

**Comment by calmnesss** [comment]
Integrated. Will take time before it get deployed. Closing as for now. Will reopen if reproduced even with changes.

**Comment by calmnesss** [comment]
QA team still able to see that issue. Reopening. More details to go as soon as they are available.

**Comment by calmnesss** [comment]
False alarm. Turned out to be configuration issue. Will keep an eye on it though.

**Comment by calmnesss** [comment]
Still have that issue even with ConcurrentHashMap.

Thanks to webhooks now we can see that after some time webhook for connection close is executed -> websocket session on CCS side is removed and it never comes back even though Controller is online and operational. Issue tends to grow over time affecting more and more accounts with no means to recover except Controller restart.

**Comment by mickz18** [comment]
What about the ping pong state on controller ? Is the controller still
sending WS message ?

Le dim. 30 juil. 2017 à 04:19, calmnesss <notifications@github.com> a
écrit :

**Comment by mickz18** [comment]
I certain case the remove session could remove the wrong websocket session.

A websocket close can happen after an open new websocket (eg: controller realize sooner than CCS that the connection is gone). In that case the handle close was removing the wrong websocket session but the websocket connection was still working.

**Comment by calmnesss** [comment]
Hey. Still seeing the issue. In brief:

1. PingPongTask fails on Controller
2. Controller closes connection and schedules reconnect at 30 seconds
3. Contoller connects and reports about successful connection
4. Something happens here after which CCS reports to HMS that connection has been closed.
5. No connection is available on CCS for that user (if using REST api to query for connection)

Found strange log snippet ( ping failed around 15:14:59):

```
August 25th 2017, 17:05:41.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 17:05:41 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-466] INFO WebsocketControllerCommandsResource:29 - WS close done for user : wjv86fbzv offset:9,379,258input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 17:05:41.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahzw12g-1gtxOSlvG _type:beehive _index:logstash-sv5-2017.08.25_score: -
-- | --

  | August 25th 2017, 17:05:41.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 17:05:41 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-466] INFO WebsocketControllerCommandsResource:24 - WS closing [CloseReason: code [1006], reason [null]] for user :wjv86fbzv offset:9,379,092 input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 17:05:41.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5beat.name:beehive1-eqix-sv5 beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahzw12g-1gtxOSlvE _type:beehive_index:logstash-sv5-2017.08.25 _score: -

  | August 25th 2017, 15:14:59.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 15:14:59 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-827] INFO WebsocketControllerCommandsResource:24 - WS closing [CloseReason: code [1000], reason []] for user : wjv86fbzv offset:7,015,217 input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 15:14:59.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5beat.name:beehive1-eqix-sv5 beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahvQA2g-1gtxOSBkL _type:beehive_index:logstash-sv5-2017.08.25 _score: -

  | August 25th 2017, 15:14:59.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 15:14:59 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-827] INFO WebsocketControllerCommandsResource:29 - WS close done for user : wjv86fbzv offset:7,015,383input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 15:14:59.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahvQA2g-1gtxOSBkN _type:beehive _index:logstash-sv5-2017.08.25_score: -

  | August 25th 2017, 15:14:59.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 15:14:59 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-827] INFO WebsocketControllerCommandsResource:16 - WS opening for user : wjv86fbzv offset:7,015,016input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 15:14:59.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahvQA2g-1gtxOSBj5 _type:beehive _index:logstash-sv5-2017.08.25_score: -

  | August 25th 2017, 15:14:59.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 15:14:59 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-827] INFO WebsocketControllerCommandsResource:18 - WS open done for user : wjv86fbzv offset:7,015,548input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 15:14:59.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahvQA2g-1gtxOSBki _type:beehive _index:logstash-sv5-2017.08.25_score: -

  | August 25th 2017, 14:00:18.000 | source:/srv/logs/beehive/ccs.log message:Aug 25 14:00:18 beehive1-eqix-sv5 [BeehiveControllerProxy] [http-nio-8083-exec-817] INFO WebsocketControllerCommandsResource:35 - WS handling message for user : wjv86fbzv offset:3,355,538input_type:log type:beehive tags:beats_input_codec_plain_applied hostname:beehive1-eqix-sv5 @timestamp:August 25th 2017, 14:00:18.000 application:hms @version:1 beat.hostname:beehive1-eqix-sv5 beat.name:beehive1-eqix-sv5beat.version:5.2.0 host:beehive1-eqix-sv5 _id:AV4ahkqW2g-1gtxOQwYw _type:beehive _index:logstash-sv5-2017.08.25
```

Log ordering could be messed up here since milliseconds are ignored. Somehow connection was closed and reopened almost immediately after which issue with offline account popped up.

**Comment by calmnesss** [comment]
More info from Natalya:

It is reproducible on my telo on frame node, I just unplugged the ethernet cable for some minutes, and right before "Reconnecting..." I plugged the cable back and on ccs I can see logs 

Aug 26 03:24:16 HMS_HOSTNAME [BeehiveControllerProxy] [http-nio-8083-exec-7] INFO  WebsocketControllerCommandsResource:16 - WS opening for user : 762htn6rs
Aug 26 03:24:16 HMS_HOSTNAME [BeehiveControllerProxy] [http-nio-8083-exec-7] INFO  WebsocketControllerCommandsResource:24 - WS closing [CloseReason: code [1000], reason []] for user : 762htn6rs
Aug 26 03:24:16 HMS_HOSTNAME [BeehiveControllerProxy] [http-nio-8083-exec-7] INFO  WebsocketControllerCommandsResource:29 - WS close done for user : 762htn6rs
Aug 26 03:24:16 HMS_HOSTNAME [BeehiveControllerProxy] [http-nio-8083-exec-7] INFO  WebsocketControllerCommandsResource:18 - WS open done for user : 762htn6rs

and the telo becomes offline from hms perspective, and the map of connected controllers on ccs doesn't have this particular controller. But controller still sends pings and receives pongs.
The reason is that on ccs there are two calls - close and open - at the same time, and connected controllers map deletes controller record by the time when ccs tries to send notification of opening websocket.
